### ISSUE TITLE
avcodec_flush_buffers doesn't work

### ISSUE THREAD

**Issue by rellla** [jc-kynesim/rpi-ffmpeg#96]
Hi,

i'm using test/6.0.1/main atm and want to solve the following issue in a player/output plugin. Platform is a RPI4.

Situation:
- Playing videos in forward direction works just fine
- Backwards slow motion or fast rewind is implemented in the player by sending AVPackets (only I-Frames) with a PTS in decreasing order. 

In normal forward playback it's no problem to send AVPackets until a decoded frame is ready, because the decoder sorts (and outputs) the decoded frames with increasing PTS, because we play forward. Also it's no problem, if we have more than 1 AVPackets internally buffered.
In backward playback (slow or fast rewind), sending more AVPackets is a problem, because the decoder doesn't perform the intended sorting. For mpeg2 this seems to be ok, but for h264 interlaced material, 11 AVPackets have to be sent to get the first frame, which isn't the AVPacket which was sent first (with the greatest PTS). The next one, we get with avcodec_receive_frame() is the one, which has the next PTS in forward direction...

I had similiar problems with rockchip (v4l2request), too. There it is solved be doing the following logic:
send AVPacket
try receive AVFrame
if we received a frame, flush the decoder, send the next AVPacket

My implementation is https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/blob/drm-atomic-gles/softhddev.c#L1149 atm.

This works for rockchip (and mpeg2) just fine - though i need to restart the decoder after a normal playback, because the expected AVPackets for a AVFrame could be > 1, if we had some B-Frames before. I'm getting a frame as soon as possible. With rpi4 mpeg2 there is a problem with this implementation, because the decoder wants 2 AVPacket, sends 1 AVFrame, I do the flush and the second one gets lost... Though i have to investigate the mpeg2 thing when software decoding on rockchip once more, i want to get h264 solved.

So my question is:
Is there any flag or hardcoded value in your code, which tells the h264_v4l2m2m decoder to wait for .... AVPackets before sending the first decoded frame? Could the decoder be forced to return the AVFrame as soon as it got all the needed AVPackets (we have just I-Frames of course, so 2 interlaced AVPackets should be enough...)?

I hope, i explained my problem in a way you can understand it ;) Thanks for your help.

Regards
Andreas

**Comment by jc-kynesim** [comment]
Yeah I understand the issue. Honestly I'd use the s/w decoder which will probably be a lot better behaved. To force out the packet that you want you should send an EndOfStream (i.e. a null packet), however I'm not honestly sure if that will work (it should) but I think you may need to restart the decoder afterwards which is going to take way more time than just using the s/w decode.
The other thing you could try is rewriting the SPS to have no reordering and a single entry DPB.

**Comment by rellla** [comment]
Ok thanks, I'll try a few things...

**Comment by rellla** [comment]
Hi, i did try to understand ffmpeg in the meanwhile. I did some changes in the code just to see the effects, but i couldn't get the decoder to return a frame, until the buffer was filled at least with 10 packets (for H264). As long as i feed it with less than 10 packets i get the EAGAIN with avcodec_receive_frame. I couldn't find the code, which is responsible for this. Can you give me a hint, where i could have a look?

Sending the EOS works kind of. But only if the decoder ist restarted. According to https://www.ffmpeg.org/doxygen/7.0/group__lavc__encdec.html a avcodec_flush_buffers() should be enough. Can you tell me, why this does not work here? I still get EOF from avcodec_receive_frame() after avcodec_flush_buffers(). Would be glad, if you can point me to some code ...

If rewriting the SPS would be worth it, where is the reordering or the DPB handled in the code? With rewriting SPS you mean setting sps->num_reorder_frames to 1?

Thanks for your help again.

**Comment by rellla** [comment]
https://www.linuxtv.org/downloads/v4l-dvb-apis/userspace-api/v4l/dev-decoder.html
After reading this one, i come to the conclusion, that best (and maybe only reasonable) way would be to force draining by sending a null packet. 
It's not guaranteed by design, that a frame should be available as soon as possible - if i understand correctly.
So i'd like to solve the avcodec_flush_buffers issue but can use close/open the decoder as a workaround for now.

**Comment by rellla** [comment]
It's me again ;)
I worked around my issue now by closing and opening the codec after sending an EOS. avcodec_flush_buffers doesn't work and avcodec_receive_frame still returns AVERROR_EOF.

Can you confirm, that avcodec_flush_buffers() isn't working as expected? If so, would it be possible to fix it in the code? I guess, https://github.com/jc-kynesim/rpi-ffmpeg/blob/test/7.1/main/libavcodec/v4l2_m2m_dec.c#L1285 would be the place where i have to look?

**Comment by jc-kynesim** [comment]
avcodec_flush_buffers should definitely clear EOS, my V4L2 stateful code definitely thinks it is doing so and in a little trivial testing on 5.1.6 on current PiOS it definitely seems to do so though I admit that my test was on seek & play for a while then loop so it isn't completely the same test as you have.

**Comment by jc-kynesim** [comment]
Do you have a small example of the stream you are trying to play that I can look at so I can see what latency I'd expect?

**Comment by rellla** [comment]
I can provide a stream, but i'm not sure if that is helpful. It's a simple DVB-S stream as MPEG TS container with h264 or mpeg2 DVB-S data in it.
https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/tree/WIP/rpiV02 is my latest code. This plugin gets the raw stream data from a player. In case of backward seek only I-Frames are sent for example.
The latest commit message tries to explain my problem.
https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/blob/WIP/rpiV02/softhddev.c#L1097 is the decoding thread loop and CodecVideoFlushBuffers https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/blob/WIP/rpiV02/codec.c#L431 should just be the avcodec_flush_buffers in my understanding - not a close/reopen.

The code works fine on a pi now, but i will retest it again with avcode_flush_buffers in CodecVideoFlushBuffers, just to confirm it not working. Rockchip with v4l2request did fine with avcodec_flush_buffers iirc.

I'm on test/7.1/main btw...

**Comment by rellla** [comment]
I changed the code to use avcodec_flush_buffers after sending EOS and this is the log:
https://paste.debian.net/1346411/
Decoding workflow is https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/blob/WIP/rpiV04/softhddev.c#L1098

Please don't bother about some additional ffmpeg logs, i added them in the code.
The workflow is as follows:

- initiate backward trickspeed
- avcodec_send_packet(ctx, packet)
- if (success) avcodec_send_packet(ctx, NULL)
- avcodec_receive_frame(ctx, &frame)
- send it to the render ringbuffer
- try again avcodec_receive_frame(ctx, &frame)
- if (AVERROR_EOF) avcodec_flush_buffers(ctx)
- try next avcodec_send_packet(ctx, packet)

After avcodec_flush_buffers i can send a new packet and force the eos again, but avcodec_receive_frame doesn't receive the frame but still reports EOF.

Is there an error in my logic, or is sth wrong with resetting EOF?

Thanks for your help again.

**Comment by rellla** [comment]
I switched back to avcodec_flush_buffers and this is the log:
https://paste.debian.net/plain/1346418

Don't bother about the additional logs - i added some more...

Based on the following decoding logic:
https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/blob/WIP/rpiV04/softhddev.c#L1098

The workflow is as follows:

- initiate backward trickspeed (player sends I-Frames in descending PTS order)
- avcodec_send_packet(ctx, packet)
- if (success) then avcodec_send_packet(ctx, NULL)
- avcodec_receive_frame(ctx, &frame)
- if (success) then send frame to render ringbuffer
- try avcodec_receive_frame again
- if (AVERROR_EOF) then avcodec_flush_buffers(ctx)
- continue with sending the next packet

The first avcodec_receive_frame returns a frame, the second one returns EOF (as expected).
After the flush i can send an AVPacket again, but avcodec_receive_frame still returns AVERROR_EOF

Is there some issue in my decoding logic, or is this an issue in avcodec_flush_buffers? 
Thanks for your help again.

Btw, it's h264 with h264_v4l2m2m

**Comment by rellla** [comment]
Hi,
i think i found it.
It works, when timeout isn't set to "0" here. https://github.com/jc-kynesim/rpi-ffmpeg/blob/test/7.1/main/libavcodec/v4l2_m2m_dec.c#L657

Either the "0" has to be higher, or prefer_dq needs to be set in my error case. I guess prefer_dq has to be set > 0 if EOS is sent. It isn't here. 

Would be cool, if you can have a look at that :)

**Comment by jc-kynesim** [comment]
If we've received an EOS (empty) packet then src_rv should be NQ_DRAINING
which should give a decent length timeout irrespective of prefer_dq in that
statement.

**Comment by rellla** [comment]
https://pastebin.com/zkmG7Mst

I will write some thoughts later, but this is my log. Lines 463 and above should be the interesting ones.

**Comment by jc-kynesim** [comment]
OK - I'm not entirely sure what is going on here but we seem to be getting
EAGAINs after the EOF - I didn't think that could happen until the status
was cleared.
Try copying the block:

    if (s->draining) {
        if (s->buf_pkt.size) {
            av_log(avctx, AV_LOG_WARNING, "Unexpected input whilst draining\n");
            av_packet_unref(&s->buf_pkt);
        }
        return NQ_DRAINING;
    }

which is outside the "if(!s->buf_pkt.size && !do_not_get)" conditional in
try_enqueue_src to just before the test "if (ret == AVERROR(EAGAIN))"
inside the condition to force EOS once we've seen it once.

**Comment by rellla** [comment]
I will try that and report.

**Comment by rellla** [comment]
Ok, so that seems to work.
Here is the new log with the same workflow and debug code: https://pastebin.com/uZApgE9x
- starting at 341 and upwards. Now we get the src_rv 4! Not tested on RPI btw, but on amlogic, but that should be equal.
Thanks for your help.

**Comment by jc-kynesim** [comment]
Hi

>rellla left a comment (jc-kynesim/rpi-ffmpeg#96)
>
>Ok, so that seems to work.
>Here is the new log with the same workflow and debug code: https://pastebin.com/uZApgE9x
>- starting at 341 and upwards. Now we get the src_rv 4! Not tested on RPI btw, but on amlogic, but that should be equal.

I was suspecting not Pi - Pi returns EOS events when requested at EOS -
what you have doesn't (which is out of spec and isn't helping here).

>Thanks for your help.

I'll add something to the effect of that patch to the tree shortly

Regards

**Comment by rellla** [comment]
I can try that on RPI and iirc, it was the same. Stay tuned ...

**Comment by rellla** [comment]
https://pastebin.com/6PkskwZt (line 2212 and upwards) is with the patch
https://pastebin.com/8NuGzjQu (line 1586 and up) is without the patch

Both seem to have the same problem.
The first frame is returned. 
Then i try a avcodec_receive_frame again, which returns EOF, which is ok.
This EOF triggers avcodec_flush_buffers.
The next avcodec_send_packet works, and the very next is triggering the drain by sending a NULL packet, which ends in a EOS request. This is also correct, i think. 

The problem comes now, because the frame could not be returned. It's around the Buffer empty...

This problem is probably a different one to amlogic, but i don't find the problem in my ffmpeg-api workflow at userspace side.

**Comment by rellla** [comment]
one nitpick regarding the logs:
`Buffer empty - reQ is_mp 1 plane.bytesused 0 bytesused 0 flag_last 1` flag_last is 0 when entering the if. i placed the log line to the end.

**Comment by rellla** [comment]
Hi @jc-kynesim ,
i did some more testing and now have a log which includes the kernel debugs:
https://gist.github.com/rellla/0ce266e0619da5a03f0ffffcc5c61b89#file-gistfile1-txt-L6128

Though userspace and kernel timestamps may not match exactly we run into some strange behaviour beginning with line 6128.

Line 6068 seems to be the eos-buffer which has a timestamp +20000 to the last. Not sure, if this is right.

May the problem be a kernelside one?

**Comment by rellla** [comment]
Hi @jc-kynesim,

i'm coming back to my issue above just popping it again and do a sum up what happens:
Hardware: RPi4
Software: [VDR*ELEC](https://github.com/Zabrimus/VDRSternELEC), which is LE-based (bleeding edge), so using rpi-ffmpeg 8.0.1, Player-App is VDR with [softhddevice-drm-gles](https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/blob/drm-atomic-gles/videostream.cpp#L375)

In trickspeed i use draining to force decoding outstanding frames. The workflow is as follows:

avcodec_send_packet(AVPacket) -> ret 0, packet sent
avcodec_send_packet(NULL) -> ret 0, drain mode forced
avcodec_receive_frame(AVFrame) -> ret 0, frame received
avcodec_receive_frame(AVFrame) -> ret AVERROR_EOF, no frame received
avcodec_flush_buffers()
avcodec_send_packet(AVPacket) -> ret 0, packet sent
avcodec_send_packet(NULL) -> ret 0, drain mode forced
avcodec_receive_frame(AVFrame) -> ret AVERROR_EOF, no frame received

The last line is the problem, because i can't get a frame decoded after i forced drain and wanted to clean up the state with avcodec_flush_buffers.

As a workaround, i do a decoder close/open instead of the flush, which works, but is a hack on userspace side.
I recently had a look into the ffmpeg code and tried some things, but i couldn't track down the problem... 
Maybe you have some new idea a few months later? Thanks for your help.

**Comment by jc-kynesim** [comment]
Thanks for the comprehensive trace.
Just to check - you are sure that the frame you give the decoder is an I-frame? (otherwise it obviously wouldn't work)
Assuming you've got that right it looks like the kernel/firmware has become confused.
@6by9 do you have anything to add here? It looks to me as if I'm doing the right thing.

**Comment by rellla** [comment]
@jc-kynesim I could check it, but im quite sure we only send I-Frames. VDR itself is responsible to send video data in trickspeed mode. And especially in backward trickspeed i already asked this question there. Answer: VDR only sends I-Frames. 

Imho we should get into trouble with opening the decoder with a Not-I-Frame in the close/open-workaround situation, too? 

I hope i do remember right, but afair the flush is working in normal playback as it should as long as we do not send a NULL-Packet to force draining. Btw, I can heal the above situation only with reopening the decoder.

Looking in the thread above, i did a parallel kernel driver log here https://github.com/jc-kynesim/rpi-ffmpeg/issues/96#issuecomment-2687735996 - the ["driver bug"](https://gist.github.com/rellla/0ce266e0619da5a03f0ffffcc5c61b89#file-gistfile1-txt-L6167) and call traces at minimum smell suspicious.... ;)

**Comment by 6by9** [comment]
I don't believe I ever implemented V4L2_DEC_CMD_START, so after EOS the only way to resume is via a pair of VIDIOC_STREAMOFF() and VIDIOC_STREAMON() calls to reset the internal EOS state of the firmware.

https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/dev-decoder.html#end-of-stream
> If the decoder encounters an end of stream marking in the stream, the decoder will initiate the Drain sequence, which the client must handle as described above, skipping the initial [VIDIOC_DECODER_CMD()](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-decoder-cmd.html#c.V4L.VIDIOC_DECODER_CMD).

https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/dev-decoder.html#drain
> Once all the OUTPUT buffers queued before the V4L2_DEC_CMD_STOP call are dequeued and the last CAPTURE buffer is dequeued, the decoder is stopped and it will accept, but not process, any newly queued OUTPUT buffers until the client issues any of the following operations:
>
>   V4L2_DEC_CMD_START - the decoder will not be reset and will resume operation normally, with all the state from before the drain,
>
>    a pair of [VIDIOC_STREAMOFF()](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-streamon.html#c.V4L.VIDIOC_STREAMOFF) and [VIDIOC_STREAMON()](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-streamon.html#c.V4L.VIDIOC_STREAMON) on the CAPTURE queue - the decoder will resume the operation normally, however any CAPTURE buffers still in the queue will be returned to the client,
>
>    a pair of [VIDIOC_STREAMOFF()](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-streamon.html#c.V4L.VIDIOC_STREAMOFF) and [VIDIOC_STREAMON()](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-streamon.html#c.V4L.VIDIOC_STREAMON) on the OUTPUT queue - any pending source buffers will be returned to the client and the Seek sequence will be triggered.

It sounds like you're trying to use option 1 which is the unimplemented one. I vaguely recall looking at implementing it ages back, and it wasn't straightforward within the firmware.

My memory is hazy over exactly the mode of operation that the stateful decoder uses within the firmware with regard buffers containing full frames and kicking the hardware accordingly.

(Apologies for the update - hit the "comment" button early).

**Comment by rellla** [comment]
Afaics, the necessary flush in the draining is done with off/on, but it seems there is only done a STREAMOFF on the output side:
https://github.com/jc-kynesim/rpi-ffmpeg/blob/test/8.0.1/main/libavcodec/v4l2_m2m_dec.c#L1357

Is it necessary to keep some order in off/on? Is the missing STREAMON related at all? It's done lazily later iirc.
I can't interpret the log: https://gist.github.com/rellla/0ce266e0619da5a03f0ffffcc5c61b89#file-gistfile1-txt-L6128 
What happens kernelside here between the STREAMOFF-STREAMON on the capture side?
Is it necessary to have the output side ON here?

**Comment by jc-kynesim** [comment]
Pretty sure that logic (as per the comments around it) is benign. Doc is here: https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/dev-decoder.html If you think I'm violating what it says then please tell me and I'll fix it. The exception to this is that I should be able to get away with V4L2_DEC_CMD_START to restart after an event where I haven't had to change the OUTPUT buffers but due to a lack of functionality in Pis firmware drivers I do the heavier weight stream off / stream on pair. (I could quirk/should the Pi for this behaviour but I don't have a "good" implementation to check the "proper" code against.)

**Comment by rellla** [comment]
> If you think I'm violating what it says then please tell me and I'll fix it.

Hi, if this is related to my answer above, i'm not sure if or what is wrong either. I just came across it and noticed, that avcodec_flush_buffers is not doing what it should according to ffmpeg docs :)

**Comment by jc-kynesim** [comment]
I think I'm doing the right thing and the kernel/firmware is getting confused somehow - it wouldn't be the first time.

As an aside, when talking about a V4L2 OUTPUT please (for my benefit) capitalize it so I am sure we are talking about the same thing, the V4L2 input=OUTPUT, output=CAPTURE thing is just so confusing. I know ffmpeg uses _output and I can't help that (in code I've written I try to use _src and _dst to avoid confusion). 

But thank you for the report - I'm just not sure how to fix the problem.
